### PR TITLE
BUG: Resolve skipfooter problem referenced in issue #7697

### LINKED ISSUE #7697: skiprows in read_fwf: Link to Issue #2065
Four years ago in issue #2065 an issue with skiprows was raised. At that time dask.dataframe.read_csv would skip the number of rows specified in skiprows from every partition, not just the first partition. This was acknowledged as a bug and fixed in pull request #2165.
https://github.com/dask/dask/pull/2165

There have been significant changes in that part of the code base and it appears that this bug has reverted.

    import dask.dataframe as dd
    import pandas as pd
    import dask
    
    print(dask.__version__)
    #  '2021.04.1'
    df = pd.read_fwf('data.txt', colspecs=<colspecs>, skiprows=1)
    ddf = dd.read_fwf('data.txt', colspecs=<colspecs>, skiprows=1).compute()
    print(len(df))
    # 1000000
    print(len(ddf))
    # 9999974
    print(ddf.npartitions)
    # 27

### PR DESCRIPTION
**Pull request by rhjmoore**
A test_skipfooter test is added in test_csv.py. This test needs to force engine='python' to suppress warnings that would otherwise trigger test failure.
Functionality added in io.csv.py to detect the last block and only include skipfooter that block's pandas read.

- [ Y] Closes #7697 
- [ Y] Tests added / passed
- [ ~] Passes `black dask` / `flake8 dask` / `isort dask` - as much as the rest of dask to be honest...

### CODE REVIEW

**Comment by jsignell** [comment]
Sorry for letting this linger so long. If you merge main into your fork the tests that are failing should pass. To get linting working properly, I recommend using [pre-commit](https://docs.dask.org/en/latest/develop.html#code-formatting)

**Comment by rhjmoore** [comment]
When I run:

    black dask
    flake8 dask
    isort dask

I get reformatting of just about every file in dask, including lots of file I haven't touched. e.g.:

    return pq.ParquetFile(fil).read(
        columns=columns,
        use_threads=False,
        use_pandas_metadata=True,
    )

becomes:

    return pq.ParquetFile(fil).read(
        columns=columns, use_threads=False, use_pandas_metadata=True,
    )

Is this right?!

**Comment by jsignell** [comment]
I don't see any issues or changes when running black, flake8 and isort on main. It seems like you might have different config somehow?

**Comment by jrbourbeau** [comment]
Yeah, there may be a mismatch between the `black`, `flake8`, etc. versions you're using locally, and those which are run in CI. To avoid these version mismatches, I'd recommend using `pre-commit` (since that's what is used in CI). To do this I would go through the following steps:

1. Merge the current `main` branch of `dask` into your branch to make sure everything is up-to-date
2. Install the `pre-commit` package with `pip install pre-commit`
3. Set up Dask's `pre-commit` hooks with `pre-commit install`
4. Run the hooks with `pre-commit run --all-files`

This should resolve any formatting issues and only impact files which you've modified as part of this PR. Hope that helps @rhjmoore -- let me know if you have any follow-up questions

File: dask/dataframe/io/csv.py [hunk 1]
```diff
@@ -105,6 +105,9 @@ def __call__(self, part):
         if not is_first:
             write_header = True
             rest_kwargs.pop("skiprows", None)
+        if not is_last:
+            write_header = True
```
Comment (jsignell) [inline dask/dataframe/io/csv.py:109]: I reran your original bug report and realized that if you include both skipfooter and skiprows the len of the output dataframe isn't correct. 

I think this is what is causing the issue:

```suggestion
```

**Comment by jsignell** [comment]
If you merge in main again, you can expect tests to pass.

File: dask/dataframe/io/csv.py (hunk 1 above)
Comment (ncclementi) [reply dask/dataframe/io/csv.py:109]: @rhjmoore  checking in here. Would you mind including this change, and merge the current main branch of dask into your branch to make sure everything is up-to-date, and will see what's the status of the tests after you push. 

Let us know if you have any troubles.

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: dask/dataframe/io/csv.py (hunk 1 above)
Comment (rhjmoore) [reply dask/dataframe/io/csv.py]: Thanks, I think you're right and I've fixed this now. Have remerged with upstream now.

**Comment by jsignell** [comment]
@dask/gpu can you trigger GPU CI?

**Comment by charlesbluca** [comment]
ok to test

**Comment by jsignell** [comment]
Thanks @rhjmoore! This is in :)

**Pull request merged**